[IMP] web: Dropdowns menus unification

BastienFafchamps opened the pull request:
We currently have too many different implementation of the same concern: an input with a dropdown/popover/select menu that can be customized, display a list of elements, let the user navigate with the keyboard, can be closed with escape, can be closed with an outside click. Also, the current navigation systems does not handle inputs properly.

This commit addresses that by doing multiple things:
- It tries to improve the Dropdown component with a cleaner props API while keeping its ease of use.
- The navigation has been improved to support inputs and better customization.
- The Dropdown now uses the popover/overlay service which will cause less z-index issues. 

---

### IMPORTANT
Using the popover service now means the div opened by the dropdown is contained in the main_components container **at the bottom of the DOM** (and thus on top of the page). This means that a lot of css selectors were failing, selectors in tests and tours have been fixed but as for CSS, we only fixed issues we encountered during development.

CSS like the one below does not work anymore,
```css
.a-cool-view {
  .dropdown-toggle {
    .dropdown-item {
      ...
    }
  }
}
```

instead it needs to be written roughly like this:
```css
.a-cool-view {
  // styles for elements in the view
  .dropdown-toggle {
    ...
  }
}

.o-dropdown--menu.dropdown-menuclass-prop {
    // styles for elements in the dropdown's menu
  .dropdown-item.dropdownitem-class-prop {
  }
}
```
You may need to use either the `menuClass` prop of the Dropdown or the `class` prop of the DropdownItem to have more specific selectors.

---

*Following tasks will convert some of the current components (like Autocomplete) to make use of the new Dropdown component.*

Project task: [3266145](https://www.odoo.com/mail/view?model=project.task&res_id=3266145&access_token=7a09516e-1a2d-4970-a909-0f5099588055)
Enterprise PR: https://github.com/odoo/enterprise/pull/48452

I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/137691).

BastienFafchamps commented:
## Dropdown:

- Simplified props API.
- Improved navigation.
- Now uses the popover and overlay services.
- The dropdown can now easly be opened and closed externally.
- The toggle does not have any default html anymore, a complete element needs to be given to the toggler slot, allowing for easier customization of the toggler element.

Basic usage of the dropdown should now look like this:

```xml
<Dropdown>
  <button class="btn btn-primary o-dropdown-caret">Options</button>
  <t t-set-slot="content">
    <DropdownItem onSelected="select()">Option A</DropdownItem>
    <DropdownItem onSelected="select()">Option B</DropdownItem>
    <DropdownItem onSelected="select()">Option C</DropdownItem>
  </t>
</Dropdown>
```

## Navigation:
- New globally available navigation hook
- Allows inputs while still being able to navigate through the items
- Has options to easily handle more specific use cases

---

## Dropdown Props API Changes

### Removed Props

togglerClass: { type: String, optional: true },

tooltip: { type: String, optional: true },

title: { type: String, optional: true },

hotkey: { type: String, optional: true },

disabled: { type: Boolean, optional: true },

skipTogglerTabbing: { type: Boolean, optional: true }

- These props can be handled by adding the attribute on the sibling node, which is used as the toggler

class: [type: String, optional: true]

- We don't need this prop anymore as the DOM is now simplified, we went from [dropdown_div, button, menu_div] to just [button, menu_div]. (This props was used on the dropdown_div)

toggler: { type: String, optional: true }

- No need to specify to use the parent as toggler as any html element can be used as a toggler for the dropdown 

startOpen: { type: Boolean, optional: true }

- This prop has no occurence in the code-base apart from tests and if ever needed it can be done using the "state" prop

menuDisplay: { type: String, optional: true, }

- This is used to set the css display property on the menu, this should be done via a class

onScroll: { type: Function, optional: true },

- This prop was too specific and can now be implemented by using the "menuRef" prop and adding the event manually

showCaret: { type: Boolean, optional: true },

- As the toggle is not longer part of the dropdown's template, it can now be added using the "o-dropdown-caret" class on the toggle.

container: { type: [Element, Function], optional: true },

- This is an option for the popover's position hook, it was used to fix some z-index issues but it is now unnecessary as the popover service fixes these issues.

autoOpen: { type: Boolean, optional: true },

- enables grouping, it is now handled by using the DropdownGroup Component.

### Added Props

state: { type: Object, shape : { isOpen: Boolean, open: Function, close: Function } }

- Allows controlling the Dropdown from the parent.

menuRef: { type: Function }

- Allows to get the ref of the dropdown's menu

options: { type: Object }

- Allows customizing the options of the popover and navigation hooks

seb-odoo commented on addons/mail/static/tests/discuss_app/discuss_tests.js:
```diff
@@ -657,47 +657,41 @@ QUnit.test("Unfollow message", async function () {
-        ],
+        contains: [[".o-mail-Message-header small", { text: "on Thread followed" }]],
     });
+    await contains(".o-mail-Message-moreMenu", { count: 1 });
```
Hello, the idea of the PR looks good, just a quick comment as I'm passing by. 
Why splitting the contains here? The goal was to check all the parts are present on the same parent at the same time, now this guarantee is removed here. Test probably won't pass as there are multiple more menu in DOM anyway.

BastienFafchamps replied:
The dropdown's content is now at the bottom of the DOM, I've added further explanations in the PR's main comment. Some elements were thus not contained in the selected parent anymore but I agree that some selectors should be more specific, I'll try to improve them.

seb-odoo commented on addons/mail/static/tests/message/message_tests.js:
```diff
@@ -12,6 +12,17 @@ import { click, contains, insertText } from "@web/../tests/utils";
 
 const { DateTime } = luxon;
 
+const SELECTORS = {
+    // Actions
+    expand: ".o-mail-Message [title='Expand']",
+    reply: ".o-mail-Message [title='Reply']",
+    markAsTodo: ".o-mail-Message [title='Mark as Todo']",
+    // Expanded Actions
+    edit: ".o-mail-Message-actions [title='Edit']",
+    delete: ".o-mail-Message-actions [title='Delete']",
+    markAsUnread: ".o-mail-Message-actions [title='Mark as Unread']",
+};
```
Please don't make indirection like that, it makes the test much harder to read/maintain. It's just as easy doing a replace all.

BastienFafchamps replied:
Indeed it was not really necessary to make this change, fixed it!

seb-odoo commented on addons/mail/static/tests/messaging_menu/messaging_menu_tests.js:
```diff
@@ -546,6 +546,7 @@ QUnit.test("basic rendering", async (assert) => {
     await contains("button", { text: "New Message" });
     await contains(".o-mail-MessagingMenu div.text-muted", { text: "No conversation yet..." });
     await click(".o_menu_systray .dropdown-toggle:has(i[aria-label='Messages'])");
+    await nextTick();
```
The following assert should be adapted to use contains rather than introducing an arbitrary tick.

seb-odoo commented on addons/mail/static/tests/tours/discuss_public_tour.js:
```diff
@@ -78,7 +78,7 @@ registry.category("web_tour.tours").add("mail/static/tests/tours/discuss_public_
         {
             content: "Click on edit",
-            trigger: ".o-mail-Message [title='Edit']",
+            trigger: ".dropdown-menu .dropdown-item[title='Edit'], .o-mail-Message [title='Edit']",
```
message prefix was helpful the understand the scope of the button, is it possible to keep it the selector functionally meaningful here? `.o-mail-Message-actions [title='Edit']` like in qunit?

robodoo commented:
Merge method set to merge directly, using the PR as merge commit message.

aab-odoo commented on addons/im_livechat/static/src/embed/frontend/overlay_container_patch.xml:
```diff
@@ -1,8 +1,8 @@
 <?xml version="1.0" encoding="UTF-8"?>
```
first commit: add explanations

aab-odoo commented on addons/web/static/src/core/utils/classname.js:
```diff
@@ -0,0 +1,71 @@
+}
+
+/**
+ * Returns an object from a class definition, wether it
```
spotted while scrolling through the code, tiny detail: whether

aab-odoo commented on addons/web/static/src/core/utils/classname.js:
```diff
@@ -0,0 +1,71 @@
+/**
```
not a big fan of these utils at first sight

aab-odoo commented on addons/web/static/tests/legacy/helpers/mountInFixture.js:
```diff
@@ -0,0 +1,103 @@
+/** @odoo-module alias=@web/../tests/helpers/mountInFixture default=false**/
```
we do not camelcase file names by convention

aab-odoo commented on addons/base_automation/static/tests/kanban_header_patch_tests.js:
```diff
@@ -15,7 +15,7 @@ function getColumn(groupIndex = 0, ignoreFolded = false) {
 async function toggleColumnActions(columnIndex) {
```
why 2 commits for usages conversion?

aab-odoo commented on addons/web/static/src/core/registry_hook.js:
```diff
@@ -0,0 +1,28 @@
+export function useRegistry(registry) {
+    const state = useState({ entries: registry.getEntries() });
+    useBus(registry, "UPDATE", ({ detail }) => {
+        const index = state.entries.findIndex(([k]) => k === detail.key);
+        if (detail.operation === "add" && index === -1) {
+            // push the new entry at the right place
+            const newEntries = registry.getEntries();
+            const newEntry = newEntries.find(([k]) => k === detail.key);
+            const newIndex = newEntries.indexOf(newEntry);
+            if (newIndex === newEntries.length - 1) {
+                state.entries.push(newEntry);
+            } else {
+                state.entries.splice(newIndex, 0, newEntry);
+            }
+        } else if (detail.operation === "delete" && index >= 0) {
+            state.entries.splice(index, 1);
+        }
+    });
```
if there's an order issue, shouldn't it be fixed in the registry directly?

aab-odoo replied:
Do you mean that in the faulty test, the operation on the registry isn't 'delete' nor 'add'? So this is only about filtering out some operations? I agree to address this post merge

aab-odoo commented:
robodoo rebase-merge r+

robodoo commented:
Merge method set to rebase and merge, using the PR as merge commit message.

robodoo commented:
@BastienFafchamps @aab-odoo linked pull request(s) odoo/design-themes#765 not ready. Linked PRs are not staged until all of them are ready.
